partner_category_unique_per_organization/models/partner.py

```diff
@@ -0,0 +1,33 @@
+# Copyright 2023 Akretion (https://www.akretion.com).
+# @author Pierrick Brun <pierrick.brun@akretion.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, models
+from odoo.exceptions import UserError
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    @api.constrains("category_id")
+    def _check_category_unique_per_organization(self):
+        for record in self:
+            for categ in record.category_id.filtered(
+                lambda categ: categ.unique_per_organization
+            ):
+                already_existing_contacts = (
+                    record.commercial_partner_id.child_ids.filtered(
+                        lambda c: c.id != record.id and categ.id in c.category_id.ids
```

Comment:
Why you use categ.id in c.category_id.ids rather categ == c.category_id ?

Reply:
> Why you use categ.id in c.category_id.ids rather than categ == c.category_id ?

`category_id`is (sadly) a [m2m](https://github.com/odoo/odoo/blob/14.0/odoo/addons/base/models/res_partner.py#L180) so need to use "in".

What is not useful is the use of the ".id"... and it might be better to use the letter "p" as "partner" rather than "c" which can be slightly confusing IMO :
`lambda p: p != record and categ in p.category_id`
... but not really a blocking question.

> Is'nt test without "for loop" possible ?

Your proposition doesn't work because of category_id being a m2m field.
But I think the for loop is necessary as you can have many categories "unique per organization" for the same organization and you want to catch which one is already in another partner, and what is the name of this partner.

Reply:
Thanks for anwering @clementmbr